pkg/extensions/sync/remote.go

```diff
@@ -5,192 +5,381 @@ package sync
+	if err != nil {
+		return nil, descriptor.Descriptor{}, err
 	}
 
-	return repos, nil
+	buf, err := man.MarshalJSON()
+
+	return buf, man.GetDescriptor(), err
 }
 
-func (registry *RemoteRegistry) GetDockerRemoteRepo(repo string) string {
-	dockerNamespace := "library"
-	dockerRegistry := "docker.io"
+func (registry *RemoteRegistry) GetOCIManifest(ctx context.Context, repo, reference string,
+) ([]byte, ispec.Descriptor, bool, error) {
+	var isConverted bool
+
+	var buf []byte
 
-	remoteHost := registry.client.GetHostname()
+	var desc ispec.Descriptor
 
-	repoRef, err := parseRepositoryReference(fmt.Sprintf("%s/%s", remoteHost, repo))
+	man, imageReference, err := registry.getManifest(ctx, repo, reference)
 	if err != nil {
-		return repo
+		return buf, desc, isConverted, err
 	}
 
-	if !strings.Contains(repo, dockerNamespace) &&
-		strings.Contains(repoRef.String(), dockerNamespace) &&
-		strings.Contains(repoRef.String(), dockerRegistry) {
-		return fmt.Sprintf("%s/%s", dockerNamespace, repo)
+	switch man.GetDescriptor().MediaType {
+	case mediatype.Docker2Manifest:
+		buf, desc, err = convertDockerManifestToOCI(ctx, man, imageReference, registry.client)
+		isConverted = true
+	case mediatype.Docker2ManifestList:
+		seen := []godigest.Digest{}
+
+		buf, desc, err = convertDockerListToOCI(ctx, man, imageReference, seen, registry.client)
+		isConverted = true
+	case mediatype.OCI1Manifest, mediatype.OCI1ManifestList:
+		buf, err = man.MarshalJSON()
+		desc = toOCIDescriptor(man.GetDescriptor())
+	default:
+		return nil, desc, false, zerr.ErrMediaTypeNotSupported
 	}
 
-	return repo
+	return buf, desc, isConverted, err
 }
 
-func (registry *RemoteRegistry) GetImageReference(repo, reference string) (types.ImageReference, error) {
-	remoteHost := registry.client.GetHostname()
+func (registry *RemoteRegistry) GetTags(ctx context.Context, repo string) ([]string, error) {
+	repoRefPath := fmt.Sprintf("%s/%s", registry.primaryHost, repo)
 
-	repoRef, err := parseRepositoryReference(fmt.Sprintf("%s/%s", remoteHost, repo))
+	repoReference, err := ref.New(repoRefPath)
 	if err != nil {
-		registry.log.Error().Str("errorType", common.TypeOf(err)).Str("repo", repo).
-			Str("reference", reference).Str("remote", remoteHost).
-			Err(err).Msg("couldn't parse repository reference")
+		return []string{}, err
+	}
 
-		return nil, err
+	tl, err := registry.client.TagList(ctx, repoReference)
+	if err != nil {
+		return []string{}, err
 	}
 
-	var namedRepoRef dockerReference.Named
+	return tl.GetTags()
+}
 
-	digest, ok := parseReference(reference)
-	if ok {
-		namedRepoRef, err = dockerReference.WithDigest(repoRef, digest)
+func convertDockerListToOCI(ctx context.Context, man manifest.Manifest, imageReference ref.Ref, seen []godigest.Digest,
+	regclient *regclient.RegClient,
+) (
+	[]byte, ispec.Descriptor, error,
+) {
+	var index ispec.Index
+
+	// seen
+	if common.Contains(seen, man.GetDescriptor().Digest) {
```

Comment:
Can we get better coverage of this?

Reply:
Actually i changed back, I don't see how you can upload docker indexes/manifests in loop.